### PR TITLE
[docs] Add details on improving training speed

### PR DESCRIPTION
**Pull request by jameslamb**
This PR proposes a large update to the documentation on tuning hyperparameters (https://lightgbm.readthedocs.io/en/latest/Parameters-Tuning.html). Today, that section contains simple suggestions like "use small `max_depth`". In this PR, I'm proposing adding some context to each recommendation so users can understand more of the intuition for *how* these parameters affect training or dataset construction. I think these descriptions can also help users to better understand how LightGBM works generally.

image

### Notes for Reviewers

* I didn't want this PR to get too big so I just covered the section `For Faster Speed`. If reviewers like this contribution, I'd be happy to add similar contributions like it in future PRs for `For Better Accuracy` and `Deal with Overfitting`.
* @StrikerRUS I pushed this as a LightGBM doc so we can test readthedocs builds. Can you please enable builds for this branch?
* Similar to #3618 , I thought about this from reading Stack Overflow answers on the `[lightgbm]`. I might have a few more PRs in the coming weeks with more documentation change proposals like this.

### CODE REVIEW

File: docs/Parameters-Tuning.rst [hunk 1]
```diff
@@ -36,15 +36,143 @@ To get good results using a leaf-wise tree, these are some important parameters:
...
+Increase ``min_data_in_bin`` to reduce training time.
+
+Decrease ``feature_fraction``
+*****************************
+
+By default, LightGBM considers all features in a ``Dataset`` during the training process. This behavior can be changed by setting ``feature_fraction`` to a value ``> 0`` and ``<= 1.0``. Setting ``feature_fraction`` to ``0.5``, for example, tells LightGBM to randomly select ``50%`` of features at the beginning of constructing each tree. This reduces the total number of splits that have to be evaluated to add each tree node.
+
+Decrease ``feature_fraction`` to reduce training time.
+
+Decrease ``max_cat_threshold``
```
Comment (jameslamb) [inline docs/Parameters-Tuning.rst:152]: I based this paragraph on the explanations in https://github.com/microsoft/LightGBM/issues/2261, https://github.com/microsoft/LightGBM/issues/2411, and https://github.com/microsoft/LightGBM/issues/2919

**Comment by StrikerRUS** [comment]
> Can you please enable builds for this branch?

Done!

File: docs/Parameters-Tuning.rst [hunk 2]
```diff
@@ -36,15 +36,143 @@ To get good results using a leaf-wise tree, these are some important parameters:
...
+'''''''''''''''''''''''''''''''''''''
+
+You might find that training is faster using a GPU-enabled build of LightGBM. See the `GPU Tutorial <./GPU-Tutorial.rst>`__ for details.
+
+Grow Shallower Trees
+''''''''''''''''''''
+
+The total training time for LightGBM increases with the total number of tree nodes added. LightGBM comes with several parameters that can be used to control the number of nodes per tree.
+
+The suggestions below will speed up training, but might hurt statistical performance.
```
Comment (btrotta) [inline docs/Parameters-Tuning.rst:58]: Maybe replace "statistical performance" with "training accuracy"? "Performance" is often used to mean "speed" so I think this would be less ambiguous. Also it might be good to mention that these options will also reduce overfitting, which can improve accuracy on the validation set.

File: docs/Parameters-Tuning.rst [hunk 3]
```diff
@@ -36,15 +36,143 @@ To get good results using a leaf-wise tree, these are some important parameters:
...
+'''''''''''''''
+
+Decrease ``num_iterations``
+***************************
+
+The ``num_iterations`` parameter controls the number of boosting rounds that will be performed. Since LightGBM uses decision trees as the learners, this can also be thought of as "number of trees".
+
+Choosing the right value of ``num_iterations`` is highly dependent on the data and objective, so this parameter is often chosen from a set of possible values through hyperparameter tuning.
+
+Decrease ``num_iterations`` to reduce training time.
```
Comment (btrotta) [inline docs/Parameters-Tuning.rst:100]: It might be good to mention that `num_iterations` is often adjusted in conjunction with `learning_rate`. See, e.g. docs by @Laurae2 here: https://sites.google.com/view/lauraepp/parameters (under "Learning rate" section).

File: docs/Parameters-Tuning.rst [hunk 4]
```diff
@@ -36,15 +36,143 @@ To get good results using a leaf-wise tree, these are some important parameters:
...
+
+Use Early Stopping
+******************
+
+If early stopping is enabled, after each boosting round the model's performance is evaluated against a validation set that contains data not available to the training process. That performance is then compared to the performance as of the previous boosting round. If the model's performance fails to improve for some number of consecutive rounds, LightGBM stops the training process.
```
Comment (btrotta) [inline docs/Parameters-Tuning.rst:105]: Similar to comment above, maybe clearer to use "accuracy" instead of "performance".

**Comment by btrotta** [comment]
@jameslamb This is great! I think it will make LightGBM a lot more accessible for people new to tree-boosting algorithms. I left a few small suggestions.

File: docs/Parameters-Tuning.rst (hunk 2 above)
Comment (jameslamb) [reply docs/Parameters-Tuning.rst:58]: oh good point! Ok I like the idea to use the term "training accuracy".

> Also it might be good to mention that these options will also reduce overfitting

I actually intentionally chose NOT to say that. There is a separate section in this document on how to improve training accuracy and another on how to fight overfitting. I wanted to keep this section only focused on runtime. I think being very straightforward like that makes it easier for people to reason about the direction and size of effects from changing these parameters.

The docs at https://lightgbm.readthedocs.io/en/latest/Parameters.html do have entries under each parameter like "can be used to reduce overfitting", "can be used to speed up training", etc.

File: docs/Parameters-Tuning.rst (hunk 3 above)
Comment (jameslamb) [reply docs/Parameters-Tuning.rst:102]: good suggestion! Updated in https://github.com/microsoft/LightGBM/pull/3628/commits/d615ad46df1c07eba298aab7d37ad95f1d513b1c

Also I never noticed before that that is a Google site :laughing:

File: docs/Parameters-Tuning.rst (hunk 4 above)
Comment (jameslamb) [reply docs/Parameters-Tuning.rst]: makes sense. I just replaced all of these with "accuracy" or "training accuracy": https://github.com/microsoft/LightGBM/pull/3628/commits/d615ad46df1c07eba298aab7d37ad95f1d513b1c

**Review by btrotta — APPROVED** [review]

**Comment by StrikerRUS** [comment]
@jameslamb Unfortunately, you couldn't modify `Parameters.rst` directly. You should include your fixes in `config.h` and then run `helpers/parameter_generator.py` script. Refer to the comment in `Parameters.rst`:
> ..  List of parameters is auto generated by LightGBM\helpers\parameter_generator.py from LightGBM\include\LightGBM\config.h file.

File: docs/Parameters-Tuning.rst [hunk 5]
```diff
@@ -36,15 +36,145 @@ To get good results using a leaf-wise tree, these are some important parameters:
...
+Choosing the right value of ``num_iterations`` and ``learning_rate`` is highly dependent on the data and objective, so these parameters are often chosen from a set of possible values through hyperparameter tuning.
+
+Decrease ``num_iterations`` to reduce training time.
+
+Use Early Stopping
+******************
+
+If early stopping is enabled, after each boosting round the model's training accuracy is evaluated against a validation set that contains data not available to the training process. That accuracy is then compared to the accuracy as of the previous boosting round. If the model's accuracy fails to improve for some number of consecutive rounds, LightGBM stops the training process.
+
+That "number of consecutive rounds" is controlled by the parameter ``early_stopping_rounds``. For example, ``early_stopping_rounds=1`` says *the first time accuracy on the validation set does not improve, stop training*.
```
Comment (StrikerRUS) [inline docs/Parameters-Tuning.rst:109]:
```suggestion
That "number of consecutive rounds" is controlled by the parameter ``early_stopping_rounds``. For example, ``early_stopping_rounds=1`` says "the first time accuracy on the validation set does not improve, stop training".
```

File: docs/Parameters-Tuning.rst [hunk 6]
```diff
@@ -36,15 +36,145 @@ To get good results using a leaf-wise tree, these are some important parameters:
 For Faster Speed
 ----------------
 
--  Use bagging by setting ``bagging_fraction`` and ``bagging_freq``
+Add More Computational Resources
+''''''''''''''''''''''''''''''''
 
--  Use feature sub-sampling by setting ``feature_fraction``
+On systems where it is available, LightGBM uses OpenMP to parallelize many operations. The maximum number of threads used by LightGBM is controlled by the parameter ``num_threads``. By default, this will defer to the default behavior of OpenMP (one thread per real CPU core or the value in environment variable ``OMP_NUM_THREADS``, if it is set). For best performance, set this to the number of real CPU cores available.
```
Comment (StrikerRUS) [inline docs/Parameters-Tuning.rst:42]:
```suggestion
On systems where it is available, LightGBM uses OpenMP to parallelize many operations. The maximum number of threads used by LightGBM is controlled by the parameter ``num_threads``. By default, this will defer to the default behavior of OpenMP (one thread per real CPU core or the value in environment variable ``OMP_NUM_THREADS``, if it is set). For best performance, set this to the number of **real** CPU cores available.
```

File: docs/Parameters.rst [hunk 7]
```diff
@@ -312,7 +312,7 @@ Learning Control Parameters
 
    -  frequency for bagging
 
-   -  ``0`` means disable bagging; ``k`` means perform bagging at every ``k`` iteration
+   -  ``0`` means disable bagging; ``k`` means perform bagging at every ``k`` iteration. Every ``k``-th iteration, LightGBM will randomly select ``bagging_fraction%`` of the data to use for the next ``k`` iterations.
```
Comment (StrikerRUS) [inline docs/Parameters.rst:315]:
```suggestion
   -  ``0`` means disable bagging; ``k`` means perform bagging at every ``k`` iteration. Every ``k``-th iteration, LightGBM will randomly select ``bagging_fraction * 100 %`` of the data to use for the next ``k`` iterations
```

File: docs/Parameters.rst [hunk 8]
```diff
@@ -358,11 +360,13 @@ Learning Control Parameters
...
 
-   -  ``<= 0`` means disable
+   -  can be used to speed up training
+
+   -  ``<= 0`` means disable (the default)
 
 -  ``first_metric_only`` :raw-html:`<a id="first_metric_only" title="Permalink to this parameter" href="#first_metric_only">&#x1F517;&#xFE0E;</a>`, default = ``false``, type = bool
 
-   -  set this to ``true``, if you want to use only the first metric for early stopping
+   -  LightGBM allows you to provide multiple evaluation metrics. Set this to ``true``, if you want to use only the first metric for early stopping.
```
Comment (StrikerRUS) [inline docs/Parameters.rst:369]:
```suggestion
   -  LightGBM allows you to provide multiple evaluation metrics. Set this to ``true``, if you want to use only the first metric for early stopping
```

File: docs/Parameters.rst [hunk 9]
```diff
@@ -382,6 +386,8 @@ Learning Control Parameters
 
 -  ``min_gain_to_split`` :raw-html:`<a id="min_gain_to_split" title="Permalink to this parameter" href="#min_gain_to_split">&#x1F517;&#xFE0E;</a>`, default = ``0.0``, type = double, aliases: ``min_split_gain``, constraints: ``min_gain_to_split >= 0.0``
 
+   -  can be used to speed up training
+
    -  the minimal gain to perform split
```
Comment (StrikerRUS) [inline docs/Parameters.rst:391]: Swap these points please as we use `can be used ...` at the very end of param description.

File: docs/Parameters.rst [hunk 10]
```diff
@@ -442,7 +448,9 @@ Learning Control Parameters
 
    -  used for the categorical features
 
-   -  limit the max threshold points in categorical features
+   -  limit number of split points considered for categorical features. See `the documentation on how LightGBM finds optimal splits for categorical features <./Features.rst#optimal-split-for-categorical-features>`_ for more details.
```
Comment (StrikerRUS) [inline docs/Parameters.rst:451]:
```suggestion
   -  limit number of split points considered for categorical features. See `the documentation on how LightGBM finds optimal splits for categorical features <./Features.rst#optimal-split-for-categorical-features>`_ for more details
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
Awesome addition! It greatly contributes to #2617, I believe. Just some minor typos below:

**Comment by jameslamb** [comment]
> @jameslamb Unfortunately, you couldn't modify `Parameters.rst` directly. You should include your fixes in `config.h` and then run `helpers/parameter_generator.py` script. Refer to the comment in `Parameters.rst`:
> 
> > ..  List of parameters is auto generated by LightGBM\helpers\parameter_generator.py from LightGBM\include\LightGBM\config.h file.

ohhhh sorry I forgot this

File: docs/Parameters.rst (hunk 9 above)
Comment (jameslamb) [reply docs/Parameters.rst]: done!

**Review by StrikerRUS — APPROVED** [review]
